controlplane/internal/controller/openshiftassistedcontrolplane_controller.go

```diff
@@ -181,22 +185,119 @@ func (r *OpenshiftAssistedControlPlaneReconciler) Reconcile(ctx context.Context,
+	result := ctrl.Result{}
+	if conditions.IsTrue(oacp, controlplanev1alpha2.KubeconfigAvailableCondition) {
+		// in case upgrade is still in progress, we want to requeue, however we also want to reconcile replicas
+		result, err = r.upgradeWorkloadCluster(ctx, cluster, oacp, log, pullsecret)
+		if err != nil {
+			return result, err
+		}
 	}
+	return result, r.reconcileReplicas(ctx, oacp, cluster)
+}
 
-	if upgrade.IsUpgradeRequested(ctx, oacp) {
-		log.Info("workload cluster upgrade has been requested, starting upgrade", "current workload cluster version", oacp.Status.DistributionVersion, "new workload cluster version", oacp.Spec.DistributionVersion)
-		if err := upgrade.UpgradeWorkloadCluster(ctx, r.Client, r.WorkloadClusterClientGenerator, oacp); err != nil {
-			log.Error(err, "failed to upgrade workload cluster")
-		}
-		log.Info("workload cluster upgrade set on ClusterVersion, waiting completion")
+func (r *OpenshiftAssistedControlPlaneReconciler) upgradeWorkloadCluster(ctx context.Context, cluster *clusterv1.Cluster, oacp *controlplanev1alpha2.OpenshiftAssistedControlPlane, log logr.Logger, pullSecret []byte) (ctrl.Result, error) {
+	// Retrieve KubeConfig
+	kubeConfig, err := util.GetWorkloadKubeconfig(ctx, r.Client, cluster.Name, cluster.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	upgrader, err := r.UpgradeFactory.NewUpgrader(kubeConfig)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	isUpdateInProgress, err := upgrader.IsUpgradeInProgress(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
 	}
-	return ctrl.Result{}, r.reconcileReplicas(ctx, oacp, cluster)
+	if isUpdateInProgress {
+		log.V(logutil.DebugLevel).Info("update is in progress...")
+		conditions.MarkFalse(
+			oacp,
+			controlplanev1alpha2.UpgradeCompletedCondition,
+			controlplanev1alpha2.UpgradeInProgressReason,
+			clusterv1.ConditionSeverityInfo,
+			"upgrade to version %s in progress",
+			oacp.Spec.DistributionVersion,
+		)
+	}
+	oacp.Status.DistributionVersion, err = upgrader.GetCurrentVersion(ctx)
+	if err != nil {
+		log.V(logutil.WarningLevel).Info("failed to get OpenShift version from ClusterVersion", "error", err.Error())
+	}
+
+	// TODO: check for upgrade errors, mark relevant conditions
+	isDesiredVersionUpdated, err := upgrader.IsDesiredVersionUpdated(ctx, oacp.Spec.DistributionVersion)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if isDesiredVersionUpdated && isUpdateInProgress {
+		log.V(logutil.WarningLevel).Info("desired version is updated, but did not complete upgrade yet. Re-reconciling")
+		return ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 1 * time.Minute,
+		}, nil
+	}
+
+	if isWorkloadClusterRunningDesiredVersion(oacp) && !isUpdateInProgress {
+		log.V(logutil.WarningLevel).Info("Cluster is now running expected version, upgraded completed")
+
+		conditions.MarkTrue(oacp, controlplanev1alpha2.UpgradeCompletedCondition)
+		return ctrl.Result{}, nil
+	}
```

Comment:
The `isWorkloadClusterRunningDesiredVersion` function https://github.com/openshift-assisted/cluster-api-agent/pull/83/files#diff-12f36ba5c9185bd1b2b7d0fc80636925dbd3f20b2509055100d988ebf99e0810R298

Doesn't tell us whether the requested version from `spec.distributionVersion` is an upgrade or downgrade compared to what the workload cluster is running. Should we be taking that into account?

Reply:
IIRC OpenShift does not support downgrade, we can forward this and let CVO give us feedback about it.
What we can do is to try this scenario and make sure we do provide feedback. WDYT?